consensus/istanbul/algorithm/algorithm.go

```diff
@@ -0,0 +1,169 @@
+// Package algorithm encapsulates logic pertaining the the celo ibft consensus algorithm.
+package algorithm
+
+import (
+	"fmt"
+
+	"github.com/celo-org/celo-blockchain/common/hexutil"
+)
+
+// Value represents the value to be agreed by consensus. It's limited size
+// means that often the value here will be a hash or some similar reduction of
+// the real value to be agreed.
+type Value [32]byte
+
+func (v Value) String() string {
+	return hexutil.Encode(v[:])
+}
+
+// Type represents the different types of consensus messages.
+type Type uint8
+
+const (
+	Propose Type = iota
+	Prepare
+	Commit
+	RoundChange
+)
+
+// In returns true if the method receiver is one of the given types.
+func (t Type) In(types ...Type) bool {
+	for _, x := range types {
+		if t == x {
+			return true
+		}
+	}
+	return false
+}
+
+func (t Type) String() string {
+	switch t {
+	case Propose:
+		return "Propose"
+	case Prepare:
+		return "Prepare"
+	case Commit:
+		return "Commit"
+	case RoundChange:
+		return "RoundChange"
+	default:
+		return fmt.Sprintf("unrecognised type: %d", t)
+	}
+}
+
+// State reprsents the discreet states that the algorithm can be in.
+type State uint8
+
+// We define these states in a separate const block to the message types to
+// maintain the equivalence with the istanbul states (istanbul/core/types.go).
+// This is because the iota identifier represents Successive untyped integer
+// values within the scope of a const block.
+const (
+	AcceptRequest State = iota
+	Preprepared
+	Prepared
+	Committed
+	WaitingForNewRound
+)
+
+// Msg represents consensus messages.
+type Msg struct {
+	MsgType         Type
+	Height          uint64
+	Round           uint64
+	Val             Value
+	RoundChangeCert *Value
+}
+
+func (m *Msg) String() string {
+	return fmt.Sprintf("T: %12v H: %12d R: %4d V: %v", m.MsgType, m.Height, m.Round, m.Val)
+}
+
+// Oracle provides access to external information. Allowing us to represent
+// concepts that are relevant to the consensus algorithm without needing to
+// have compile time dependencies on all the code required to manage an compute
+// those concepts. Currently it is managing more than it should be, since
+// certain logic that is really part of the consensus algorithm is abstracted
+// away by the Oracle, for example managing the round and state of the
+// algorithm and the logic for processing round change messages. In time we
+// should be able to move this logic here and slim down the oracle's interface.
+type Oracle interface {
+	CurrentState() State
+	QuorumCommit(height, round uint64, val Value) bool
+	QuorumPrepare(height, round uint64, val Value) bool
+	DesiredRound() uint64
+	QuorumRoundChange() *uint64
```

Comment:
I'd prefer an interface more like `bool, uint64` rather than a pointer. The best interface would be `Some(u64)` (using rust syntax), but absent that, the pair makes slightly more sense to me. (I was able to figure out the intention from reading the definition, but it still feels tricky to me)

Reply:
Definitely these methods could do with some documentation describing the meaning of what they return.

I'm wondering if you could describe a bit more what seems "tricky" to you about this approach?

In this specific case I think using a bool could result in slightly neater use of these methods where they are currently used, so I'm not averse to it. In general I think representing non existence with nil makes a lot of sense, nothing else in the language indicates theres nothing there more than nil!